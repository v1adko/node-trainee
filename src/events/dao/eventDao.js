const BaseDao = require('./baseDaoMongoose');
const Event = require('../models/event');
const coordinationsService = require('../services/coordinations');

class EventDao extends BaseDao {
  create(item) {
    this.checkType(item);
    return coordinationsService.set(item)
      .then(chengedItem => super.create(chengedItem));
  }
}

module.exports = new EventDao(Event);
