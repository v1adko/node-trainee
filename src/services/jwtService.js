import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
