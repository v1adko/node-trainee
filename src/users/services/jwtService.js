const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
