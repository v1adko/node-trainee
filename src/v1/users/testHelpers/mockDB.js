import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
