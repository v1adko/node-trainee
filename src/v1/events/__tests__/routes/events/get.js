import simulate from '../../../../../utils/tests-utils/request-helper';
import jwtService from '../../../../services/jwt-service';
import mockDB from '../../../../../utils/tests-utils/mock-db';
import UserFields from '../../../../../utils/tests-utils/test-user-fields';
import eventDao from '../../../event-dao';

const ROUTE = '/v1/events';

const { username, password, invalidToken } = UserFields;

let user;
let userToken;

async function clean() {
  await eventDao.model.remove();
  userToken = null;
}

async function createUser() {
  user = await mockDB.createUser(username, password);
  userToken = jwtService.generateJwt(user);
}

beforeAll(async () => {
  await clean();
});
afterAll(mockDB.closeConnection);

describe(`Test the ${ROUTE} path`, () => {
  afterAll(clean);
  beforeAll(createUser);

  it('should return all events in response on GET method', async () => {
    const postBody = { address: 'Чичибабина 1' };
    await simulate.post(ROUTE, 200, postBody, userToken);
    const body = await simulate.get(ROUTE, 200, userToken);

    expect(Object.keys(body).length).toBe(1);
  });

  it('should not return all events in response on GET method, because user token is not valid', async () => {
    const { error } = await simulate.get(ROUTE, 401, invalidToken);

    expect(error).toMatchSnapshot();
  });
});
