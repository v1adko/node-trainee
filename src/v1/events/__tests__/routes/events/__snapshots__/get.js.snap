// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/events path should not return all events in response on GET method, because user token is not valid 1`] = `
Object {
  "code": 401,
  "message": "AuthorizationError(401): Invalid token, please repeat authentication.",
  "name": "AuthorizationError",
}
`;
